### ISSUE TITLE
GPDB bounce on ERROR: attribute number 2 exceeds number of columns 1 (execQual.c:651)

### ISSUE THREAD

**Issue by kelvien** [greenplum-db/gpdb#2865]
This error sometimes shows up while it disconnects all sessions in GPDB, possibly restart GPDB.
This happens on running below query:
`SELECT * FROM test('0') WHERE name IS NOT NULL`

function test definition:
`CREATE OR REPLACE FUNCTION test(val text) RETURNS some_type AS
$BODY$
    return {
        'address': None,
        'name': None}
$BODY$
  LANGUAGE plpythonu VOLATILE;`
type some_type definition:
`CREATE TYPE some_type AS
   (address text,
    name text);`

Notice that it does not throw an error on running this query:
`SELECT * FROM test('0')`

Version: "PostgreSQL 8.2.15 (Greenplum Database 4.3.11.1 build commit: 9fad76371cc5554375f0c32ae46571eccf41c75b) on x86_64-unknown-linux-gnu, compiled by GCC gcc (GCC) 4.4.2"

**Comment by hsyuan** [comment]
@hlinnaka 
Is 2 column right here?

 address | name
---------+------
         |
(1 row)

I thought UDF test(val text) just return 1 value with type of `some_type`.

**Comment by hlinnaka** [comment]
Yeah, that 2 column result seems like the correct result here. With `SELECT test()`, you would get one value of type `some_value`, but "SELECT * FROM test()` works differently, and should return those two columns..

**Comment by hsyuan** [comment]
GPDB replaces single row returning volatile (sirv) functions with a more complicated construct so that it may be evaluated in a initplan subsequently, specifically in function `replace_sirvf_rte`. Reason for this is that this sirv function may perform DDL/DML/dispatching and it can do all this only if it is evaluated on the master as an initplan. 

But we failed to replace the target list in this issue's case.

**Comment by hsyuan** [comment]
The plan generated by GPDB is the following, instead of a simple function scan:
```
Subquery Scan test  (cost=0.26..0.28 rows=1 width=32)
   Filter: test.name IS NOT NULL
   ->  Result  (cost=0.26..0.27 rows=1 width=0)
         InitPlan
           ->  Result  (cost=0.00..0.26 rows=1 width=0)
```

**Comment by hlinnaka** [comment]
Fixed by commit b4025b4e0f.
